documentation.typ

```diff
@@ -135,6 +135,34 @@ parameters:
 )
 ")
 
+Sub-examples can be achieved by using the same `#gloss` function but surrounding each sub-example with parentheses, separated by commas. A global `header` field for the set can be added.
```

Comment:
I don't think I like trying to re-use the `gloss` function with radically-different argument structures depending on how you want a gloss formatted. Maybe this is a style the typst ecosystem is converging on, I would want to check and see what other typst packages that have complicated functions do. But even so, I think that if it would be hard to write a type signature for a function, it's probably not a great API for a user.

I'm not sure what a good alternative would be. The most straightforward thing would be to define a new function called something like `gloss-block`, that had the signature you propose for `gloss` but without changing the original `gloss` function. I'm not sure I like that either, though, since each gloss subentry is in some sense the same structure as the single entry of a gloss without subentries.

I'll keep thinking about this and maybe see if anyone in the typst chat has a suggestion about what would make for a good API. This package is on a 0.x semver version and makes no claims to stability, so there are no issues with changing the arguments of `gloss` and `numbered-gloss` if it makes sense to do.

Reply:
IMHO it is more elegant, and more straightforward and simple from the user’s perspective, not to have different functions for basically two variants of the same thing. This can be thought of like this: an example with no sub-examples is really an example with a list of sub-examples of size 1, and there is a convention where a list of sub-examples of size 1 is not marked with letters (you don’t have sub-example (a) and that’s it). It’s like a *wug* is a special case of *wugs* where the number of wugs is 1… ;-)

Originally I tried to retain the behaviour of `#gloss` for simplex examples (the only kind before this PR) as it is, but you are totally right: we are on 0.x SemVer after all, so it is acceptable to change the argument structure a bit. So, I think that for the sake of consistency of argument structure we can define `#gloss` to take each sub-example as a dictionary, even if there is only one example (i.e. not treating it as special argument-structure-wise). This way all sub-examples are consistently passed through the [argument sink](https://typst.app/docs/reference/foundations/arguments/#argument-sinks). For example₁, the example₂ from the previous PR would go like this (note the new parentheses):

```typst
#numbered-gloss(
 (
  header: [Middle Welsh; modified from _Grammatical number in Welsh_ (1999) by Silva Nurmio (§~2.1.1)],
  source: ([ac], [ny], [allvs], [y], [dewinyon], [atteb], [idav]),
  morphemes: ([and], [#neg], [be_able.#smallcaps[pret].3#sg], [#smallcaps[def]], [sorcerer.#pl], [answer.#smallcaps[inf]], [to.3#sg.#smallcaps[m]]),
  translation: [and the sorcerers could not answer him],
  label: "sorcerers",
  label-supplement: [Example]
 )
)
```

This is elegant and consistent, and presents only a tiny ‘inconvenience’ in the form of another pair of parentheses (but the way I see it is more convenient — in a deeper sense — to have all calls for the function be of the same kind).

If we agree on this way I will push an additional commit that makes use of the proposed behaviour (updating the documentation accordingly, of course).

Reply:
I'm not too bothered by the strategy of forcing the `gloss` function to always accept a list of dictionaries, where the keys of those dictionaries represent the actual meaningful arguments to the function. It's also easy enough to define a new function `single-gloss` (or something along those lines) that just unwraps one layer of dictionary for the common case of only having one example per gloss - that gets rid of the need to type the extra parens all the time unless you're using the functionality.

Still, I like the idea of there being one basic function that generates the "meat" of a gloss - some function with `source`, `morphemes`, and `translation` fields, in other words. And then this function would be optionally wrapped with other grouping functions that would have the effect of "reaching inside" and changing the indentation/numbering scheme of the generated glosses. This feels like it would make for a more ergonomic API. It's also possible that the "right" way to design this API long-term will require new scripting features added to typst that don't currently exist.

For now, I'll say go with the dictionary-list strategy you described above, and add a `single-gloss` function that unwraps the parens in the single-gloss case. That way this code can be merged and it can be useful to people now. I expect that in the future I'll want to change the argument structure of these functions, but that can happen in a future version.

Reply:
I just gave the documentation you linked for `expex` a skim. It looks like they go with the convention of `\ex` for a single example, and `\pex` for a multipart example. Perhaps a reasonable typst equivalent of this API would be functions called `gloss` and `multi-gloss` (and then `numbered-gloss` and `numbered-multi-gloss`, although I'm now wondering if defining `numbered-gloss` in the same namespace is actually a principled choice).

Reply:
> Still, I like the idea of there being one basic function that generates the "meat" of a gloss - some function with source, morphemes, and translation fields, in other words. And then this function would be optionally wrapped with other grouping functions that would have the effect of "reaching inside" and changing the indentation/numbering scheme of the generated glosses. This feels like it would make for a more ergonomic API. It's also possible that the "right" way to design this API long-term will require new scripting features added to typst that don't currently exist.

I totally agree. Moreover, I think this should be clearly communicated by the proper terms. See https://github.com/neunenak/typst-leipzig-glossing/pull/7/commits/fd5048fa6b335d6cef1a3f883d60b8aebacad91f, which separates the two entities you describe and names them `gloss()` and `example()` respectively. I think that’s the best solution for this.

---

BTW, unrelated to the matter in question: many thanks for introducing me to Radicle 👾! It’s brilliant, and I really hope it’ll catch. I for one moved [some of my repos](https://app.radicle.xyz/nodes/seed.radicle.garden/users/did:key:z6MkkshLg6RSzq59qDVN72Tmsxdq7L8iJHvu3FHULA9VkZ9u) to the distributed network. This PR is best to be completed here, but I have some new improvements I will send through Radicle after this one is done.

Reply:
> I just gave the documentation you linked for `expex` a skim. It looks like they go with the convention of `\ex` for a single example, and `\pex` for a multipart example. Perhaps a reasonable typst equivalent of this API would be functions called `gloss` and `multi-gloss` (and then `numbered-gloss` and `numbered-multi-gloss`, although I'm now wondering if defining `numbered-gloss` in the same namespace is actually a principled choice, especially if the API continues to gain variant functions).

In short, I don’t think ExPex’s API is well-designed.